# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: accessible/generic/DocAccessible.cpp
# Commit: d08313ba26df
# Full Hash: d08313ba26df045303607621d55812b37ae19cf3
# Author: James Teh <jteh@mozilla.com>
# Date: 2020-06-12 14:56:55
# Description:
#   Bug 1645067: Set the top level DocAccessibleChild as early as possible. r=MarcoZ
#   
#   Otherwise, if we're replacing a previous top level doc, we might refer to the old top level doc during initialization.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D79398
# ==============================================================================

diff -r 7c5fe3b1bec5 -r d08313ba26df accessible/generic/DocAccessible.cpp
--- a/accessible/generic/DocAccessible.cpp	Fri Jun 12 02:38:35 2020 +0000
+++ b/accessible/generic/DocAccessible.cpp	Fri Jun 12 07:11:02 2020 +0000
@@ -1614,10 +1614,18 @@
       nsIDocShell* docShell = mDocumentNode->GetDocShell();
       if (RefPtr<dom::BrowserChild> browserChild =
               dom::BrowserChild::GetFrom(docShell)) {
+        // In content processes, top level content documents are always
+        // RootAccessibles.
+        MOZ_ASSERT(IsRoot());
         DocAccessibleChild* ipcDoc = IPCDoc();
-        if (!ipcDoc) {
+        if (ipcDoc) {
+          browserChild->SetTopLevelDocAccessibleChild(ipcDoc);
+        } else {
           ipcDoc = new DocAccessibleChild(this, browserChild);
           SetIPCDoc(ipcDoc);
+          // Subsequent initialization might depend on being able to get the
+          // top level DocAccessibleChild, so set that as early as possible.
+          browserChild->SetTopLevelDocAccessibleChild(ipcDoc);
 
 #if defined(XP_WIN)
           IAccessibleHolder holder(
@@ -1633,10 +1641,6 @@
           ipcDoc->SendPDocAccessiblePlatformExtConstructor();
 #endif
         }
-
-        if (IsRoot()) {
-          browserChild->SetTopLevelDocAccessibleChild(ipcDoc);
-        }
       }
     }
   }